File: Python3/801 - 1000/803 Bricks Falling When Hit.py
# 424ms
class Solution:
    def hitBricks(self, grid, hits):
        def dfs(i, j):
            if not (0 <= i < m and 0 <= j < n) or grid[i][j] != 1: return 0
            grid[i][j] = 2
            return 1 + sum(dfs(x, y) \
                           for x, y in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)])
        
        def is_connected(i, j):
            return i == 0 or any([0 <= x < m and 0 <= y < n and grid[x][y] == 2 \
                                  for x, y in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]])
        
        m, n = len(grid), len(grid[0])
        for i, j in hits:
            grid[i][j] -= 1
        for j in range(n):
            dfs(0, j)
        res = [0] * len(hits)
        for k in reversed(range(len(hits))):
            i, j = hits[k]
            grid[i][j] += 1
            if grid[i][j] == 1 and is_connected(i, j): res[k] = dfs(i, j) - 1
            
        return res
    

File: 801 - 1000/803 Bricks Falling When Hit.cc
// 125ms
class Solution {
private:
    int m, n;
    
public:
    vector<int> hitBricks(vector<vector<int>> &grid, vector<vector<int>> &hits) {
        m = grid.size(), n = grid[0].size();
        int h = hits.size();
        auto l = [&] (int x, int y) {
            return x >= 0 && x < m && y >= 0 && y < n && grid[x][y] == 2;
        };
        
        for (auto h : hits)
            grid[h[0]][h[1]] -= 1;
        for (int j = 0; j < n; ++j)
            dfs(grid, 0, j);
        vector<int> res(h, 0);
        for (int k = h - 1; k >= 0; --k) {
            int i = hits[k][0];
            int j = hits[k][1];
            grid[i][j] += 1;
            if (grid[i][j] == 1 && 
                (i == 0 || l(i - 1, j) || l(i + 1, j) || l(i, j - 1) || l(i, j + 1)))
                res[k] = dfs(grid, i, j) - 1;
        }
        return res;
    }

private:
    int dfs(vector<vector<int>> &grid, int i, int j) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != 1)
            return 0;
        grid[i][j] = 2;
        return 1 + dfs(grid, i - 1, j) + dfs(grid, i + 1, j) + 
            dfs(grid, i, j - 1) + dfs(grid, i, j + 1);
    }
};
